runtime/startup_nat.c

```diff
@@ -77,11 +87,39 @@ static void init_static(void)
     if (caml_code_segments[i].end > caml_code_area_end)
       caml_code_area_end = caml_code_segments[i].end;
   }
-  /* Register the code in the table of code fragments */
+}
+
+/* Initialize the table of code fragments and add the code of the executable
+   as first fragment. */
+
+static void init_code_fragments(char_os * exe_name)
+{
+  struct code_fragment * cf;
+  int fd;
+
   cf = caml_stat_alloc(sizeof(struct code_fragment));
   cf->code_start = caml_code_area_start;
   cf->code_end = caml_code_area_end;
-  cf->digest_computed = 0;
+  /* Issue #5942: it is not enough to hash the code, global data
+     must also be included in the digest.  However, global data,
+     once loaded into main memory, varies from run to run of a
+     given executable, owing to ASLR.  Hence, we use a digest of
+     the whole executable file as digest for the main code fragment.
+     This digest is computed on-demand from a file descriptor
+     opened on the executable file. */
+  fd = open_os(exe_name, O_RDONLY | O_BINARY, 0);
```

Comment:
I haven’t checked the details for how the executable image is locked by Windows while running (it is locked, though), but I’m wondering if the native fd may want opening with the various sharing flags and (possibly) some additional error handling? Alternatively, given that threading is guaranteed present, the hash could be computed in parallel, but this is probably over-thinking it.

Reply:
The test passed on several Windows platforms from CI.  (A couple other platforms didn't work, but for unrelated reasons.)  So, I hope the code that reads the executable file seems to work under Windows like it does under Unix.  I'm not sure what you want to achieve with other sharing flags?

Reply:
Indeed - I was referring to the difference on Windows where by default you cannot delete or rename a file which is open (this has stung us elsewhere before - especially when antivirus programs keep files open for a long time after they’ve finished in the testsuite). My memory is that you can rename, but not delete, a running program - it seems worth checking (which I’m happy to do, just not this second!)

Reply:
I can confirm that with this PR, all Windows programs compiled with ocamlopt lose the ability to be renamed while running.

Two options (both of which I'd be happy to contribute) - specify `O_SHARE_DELETE` (which I think requires using `CreateFile` directly and then `_open_osfhandle`) or split this off into something like `caml_init_process_hash` called here which on Unix opens the fd and on Windows spawns a thread to do the hashing straight away and then later `caml_get_process_hash` which on Unix computes the hash and on Windows blocks (if necessary) until the thread finishes. Given that the process will have just been loaded by the OS (and mapped), computing the hash eagerly in parallel at the start should at least not involve (much) disk access, and the ugliness of the difference in behaviour can at least be masked by functions in `unix.c` and `win32.c`.

See [this old comment](https://github.com/ocaml/ocaml/blob/4.07/testsuite/makefiles/Makefile.several#L101-L108) from the pre-ocamltest testsuite for where this has bitten us before.

As an aside - is there enough information at link time in ocamlopt to be able to determine whether marshalling of closures is actually needed in a given program (it's correct I think that if neither Dynlink nor Marshal is actually used, then we know that this functionality is never needed)?

Reply:
I'm not sure it is that important to be able to rename or delete executable files while they are running.  But just to avoid a possible regression in this department, I pushed a new version of the PR that opens the executable file in FILE_SHARE_DELETE mode, using `CreateFile` as suggested.

I'm not enthusiastic about computing the hash speculatively in a background thread.  For one thing this would still waste a bit of resources (energy, disk or memory bandwidth) if the program doesn't use marshaling of closures.  For another thing, terminating the thread cleanly when the main program exits is not completely trivial, we ran into this problem in some of the tests for threads.